server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs

```diff
@@ -49,110 +55,129 @@ import           Hasura.Prelude
 import           Hasura.RQL.Types
 import           Hasura.SQL.Value
 
-newtype RemoteRelKey =
-  RemoteRelKey Int
-  deriving (Eq, Ord, Show, Hashable)
-
-data RemoteRelField =
-  RemoteRelField
-    { rrRemoteField   :: !RemoteField
-    , rrArguments     :: !ArgsMap
-    , rrSelSet        :: !SelSet
+-- | When parameterised by 'RRF_Tree' this is like a 'VQ.Field' but with a path
+-- indicating where its response should be spliced back into the hasura result.
+--
+-- When parameterised by 'RRF_Splice' this represents just the metadata needed to
+-- splice our remote result.
+data RemoteRelBranch (p :: RRF_P) =
+  RemoteRelBranch
+    { rrRemoteRelationship :: !RemoteRelationship
+    -- ^ The hasura to remote schema field mapping.
+    , rrArguments     :: !(Param p ArgsMap)
+    -- ^ User-provided arguments. See '_fArguments'.
+    , rrSelSet        :: !(Param p SelSet)
+    -- ^ Selection set of 'rrAlias' (may be empty)
     , rrRelFieldPath  :: !RelFieldPath
+    -- ^ Splice path.
     , rrAlias         :: !G.Alias
+    -- ^ A possible alias (provided by client) for `rtrName .
+    -- rmfRemoteRelationship . rrRemoteField`, otherwise this will be defaulted
+    -- to that value.
     , rrAliasIndex    :: !Int
+    -- ^ index of selection set at 'rrRelFieldPath' at which this remote field should be spliced.
+    -- This is morally the last component of 'rrRelFieldPath'.
     , rrPhantomFields :: ![Text]
+    -- ^ Hasura fields which are not in the selection set, but are required as
+    -- parameters to satisfy the remote join.
     }
-  deriving (Show)
-
+deriving instance Show (RemoteRelBranch 'RRF_Tree)
+deriving instance Show (RemoteRelBranch 'RRF_Splice)
+
+-- Trying out this pattern in a low-stakes setting: this lets us express
+-- variations of a data type but with less friction or boilerplate than if the
+-- parameter were a 'Functor' directly, say. Concretely e.g. code does not need
+-- to unwrap an Identity when accessing rrSelSet.
+--
+-- Related or complimentary:
+--   https://hackage.haskell.org/package/higgledy 
+--   https://reasonablypolymorphic.com/blog/higher-kinded-data/ 
+--   https://hackage.haskell.org/package/barbies
+
+-- | This could live in "Hasura.Prelude" if we want to do this elsewhere.
+type family Param (p :: k) x
+-- DataKinds:
+data RRF_P = RRF_Tree | RRF_Splice
+-- Use the special parameter to determine whether to keep or drop fields (when
+-- used just as a splice):
+type instance Param 'RRF_Tree   a = a
```

Comment:
Very interesting. So if I understand this correctly at a high level:

If `RemoteRelBranch RRF_Tree` then we will get the `arguments` and `selSet` fields 
else if `RemoteRelBranch RRF_Splice` then we will get `arguments` and `selSet` to be their unit values?

Reply:
Yep that's right

Reply:
The `Param` trick makes this one step fancier from a basic parameterized type, since we can avoid special unwrapping of the special fields (e.g. `getConst`), and it lets us make the parameter a tag that expresses the semantics so that the entire parameterized type hopefully tells us a bit more than e.g. `RemoteRelBranch (Const ())` or something